File: cs/link.py
"""`cs link`: keep each project's Claude files in sync between the side-store
(<config-repo>/projects/<name>/) and every checkout of that project.

Files are COPIED (not symlinked) because Claude Code rewrites some of them.
Newer content wins in either direction; the side-store is the source of truth
for "which files exist". `autoMemoryDirectory` is injected into the checkout's
.claude/settings.local.json at copy time and stripped on the way back, so the
side-store stays machine-independent.
"""
from __future__ import annotations

import json
import os
import shutil
from pathlib import Path
from typing import Dict, List, Optional, Set, Tuple

from . import gitutil, jsonmerge, paths
from .config import Machine
from .manifest import Manifest, Project
from .ui import act, info, ok, warn

MANAGED_ROOT_FILES = ("CLAUDE.md", "CLAUDE.local.md", ".mcp.json")
MANAGED_DIR = ".claude"
SKIP_UNDER_CLAUDE = ("worktrees", "plans", "settings.json")   # never treated as side-store content
EXCLUDE_LINES = [".claude/", ".mcp.json", "CLAUDE.md", "CLAUDE.local.md"]
SETTINGS_LOCAL = ".claude/settings.local.json"
NOT_SYNCED_SIDE_DIRS = ("memory", "secrets")


def side_store(repo: Path, p: Project) -> Path:
    return repo / "projects" / p.name


def memory_dir(repo: Path, p: Project) -> Path:
    return side_store(repo, p) / "memory"


def checkouts(p: Project, ws: Path) -> List[Path]:
    """Every directory that should carry the project's Claude files."""
    root = p.checkout_root(ws)
    if not root.exists():
        return []
    if p.kind != "git" or not gitutil.is_repo(root):
        return [root]
    wts = gitutil.worktrees(root)
    return wts or [root]


def _managed_rel_files(base: Path) -> Set[str]:
    rels: Set[str] = set()
    for f in MANAGED_ROOT_FILES:
        if (base / f).is_file():
            rels.add(f)
    cdir = base / MANAGED_DIR
    if cdir.is_dir():
        for dp, dns, fns in os.walk(cdir):
            rel_dir = Path(dp).relative_to(base)
            if len(rel_dir.parts) >= 2 and rel_dir.parts[1] in SKIP_UNDER_CLAUDE:
                dns[:] = []
                continue
            for fn in fns:
                rel = str(rel_dir / fn)
                if rel == f"{MANAGED_DIR}/settings.json":
                    continue
                rels.add(rel)
    return rels


def _side_rel_files(side: Path) -> Set[str]:
    rels: Set[str] = set()
    if not side.is_dir():
        return rels
    for dp, dns, fns in os.walk(side):
        rel_dir = Path(dp).relative_to(side)
        if rel_dir.parts and rel_dir.parts[0] in NOT_SYNCED_SIDE_DIRS:
            dns[:] = []
            continue
        for fn in fns:
            rels.add(str(rel_dir / fn) if rel_dir.parts else fn)
    return rels


def _normalize(rel: str, data: bytes) -> bytes:
    """Content as it should look in the side-store (machine-independent)."""
    if rel == SETTINGS_LOCAL:
        try:
            d = json.loads(data.decode("utf-8") or "{}")
        except (ValueError, UnicodeDecodeError):
            return data
        d.pop("autoMemoryDirectory", None)
        return jsonmerge.dumps(d).encode("utf-8") if d else b""
    return data


def _localize(rel: str, data: bytes, mem_dir: Path) -> bytes:
    """Content as it should look inside a checkout on this machine."""
    if rel == SETTINGS_LOCAL:
        try:
            d = json.loads(data.decode("utf-8") or "{}") if data.strip() else {}
        except (ValueError, UnicodeDecodeError):
            d = {}
        d["autoMemoryDirectory"] = paths.contract(mem_dir)
        return jsonmerge.dumps(d).encode("utf-8")
    return data


def _state_file(p: Project) -> Path:
    return paths.state_dir() / "link" / f"{p.name}.json"


def _load_state(p: Project) -> Set[str]:
    f = _state_file(p)
    if f.exists():
        try:
            return set(json.loads(f.read_text()).get("files", []))
        except ValueError:
            pass
    return set()


def _save_state(p: Project, files: Set[str]) -> None:
    f = _state_file(p)
    f.parent.mkdir(parents=True, exist_ok=True)
    f.write_text(json.dumps({"files": sorted(files)}, indent=2))


def _write(path: Path, data: bytes, mtime: Optional[float] = None) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_name(path.name + ".cs-tmp")
    tmp.write_bytes(data)
    if mtime is not None:
        os.utime(tmp, (mtime, mtime))
    os.replace(tmp, path)


def ensure_exclude(checkout: Path, check: bool, changes: List[str]) -> None:
    if not gitutil.is_repo(checkout) and not (checkout / ".git").is_file():
        return
    ex = gitutil.info_exclude(checkout)
    text = ex.read_text() if ex.exists() else ""
    lines = text.splitlines()
    missing = [l for l in EXCLUDE_LINES if l not in lines]
    if not missing:
        return
    changes.append(f"exclude {', '.join(missing)} in {paths.contract(checkout)}")
    if not check:
        ex.parent.mkdir(parents=True, exist_ok=True)
        block = "\n".join(missing) + "\n"
        ex.write_text(text + ("" if not text or text.endswith("\n") else "\n") + "# claude-share managed files\n" + block)


def sync_project(repo: Path, p: Project, ws: Path, check: bool = False) -> List[str]:
    """Two-way newer-wins sync for one project. Returns human-readable changes."""
    changes: List[str] = []
    side = side_store(repo, p)
    targets = checkouts(p, ws)
    if not targets:
        return changes
    mem = memory_dir(repo, p)
    if not mem.exists() and not check:
        mem.mkdir(parents=True, exist_ok=True)

    previously = _load_state(p)
    side_rels = _side_rel_files(side)
    all_rels: Set[str] = set(side_rels)
    for t in targets:
        all_rels |= _managed_rel_files(t)
    all_rels.add(SETTINGS_LOCAL)   # always exists in checkouts (memory redirect)

    final: Set[str] = set()
    for rel in sorted(all_rels):
        sp = side / rel
        side_exists = sp.is_file()
        side_data = _normalize(rel, sp.read_bytes()) if side_exists else None
        side_mtime = sp.stat().st_mtime if side_exists else -1.0

        # newest candidate among side-store and checkouts
        best_data, best_mtime, best_from = side_data, side_mtime, "side-store"
        for t in targets:
            tp = t / rel
            if tp.is_file():
                data = _normalize(rel, tp.read_bytes())
                mt = tp.stat().st_mtime
                if data != best_data and mt > best_mtime + 1e-6:
                    best_data, best_mtime, best_from = data, mt, paths.contract(t)

        deleted_in_side = (not side_exists) and rel in previously and rel != SETTINGS_LOCAL
        if deleted_in_side:
            for t in targets:
                tp = t / rel
                if tp.exists():
                    changes.append(f"remove {rel} from {paths.contract(t)} (deleted in side-store)")
                    if not check:
                        tp.unlink()
            continue

        if best_data is None:
            if rel == SETTINGS_LOCAL:
                best_data, best_mtime = b"", side_mtime
            else:
                continue

        # side-store gets the winner (settings.local.json only if non-empty)
        if best_data != side_data and (best_data or rel != SETTINGS_LOCAL):
            changes.append(f"side-store <- {rel} (from {best_from})")
            if not check:
                _write(sp, best_data, best_mtime if best_mtime > 0 else None)
        if best_data or rel != SETTINGS_LOCAL:
            final.add(rel)

        # every checkout gets the localized winner
        for t in targets:
            tp = t / rel
            want = _localize(rel, best_data, mem)
            have = tp.read_bytes() if tp.is_file() else None
            if have != want:
                changes.append(f"{paths.contract(t)}/{rel} <- side-store")
                if not check:
                    _write(tp, want, best_mtime if best_mtime > 0 else None)

    for t in targets:
        ensure_exclude(t, check, changes)
    if not check:
        _save_state(p, final)
    return changes


def run(repo: Path, m: Machine, man: Manifest, names: Optional[List[str]] = None, check: bool = False) -> int:
    ws = man.workspace(m)
    projects = [p for p in man.selected(m) if not names or p.name in names]
    if names:
        unknown = set(names) - {p.name for p in man.projects.values()}
        if unknown:
            raise SystemExit(f"cs: unknown project(s): {', '.join(sorted(unknown))}")
    total = 0
    for p in projects:
        if not checkouts(p, ws):
            continue
        changes = sync_project(repo, p, ws, check)
        for c in changes:
            (info if check else act)(f"{p.name}: {c}")
        total += len(changes)
    if total == 0:
        ok("project files in sync")
    return total

File: src/link.ts
/** cs link: project state (<share>/projects/<name>/) ⇄ every checkout, newer wins; autoMemoryDirectory injected. */
import { existsSync, mkdirSync, readdirSync, readFileSync, renameSync, rmSync, statSync, unlinkSync, utimesSync, writeFileSync } from "node:fs";
import { dirname, join, relative } from "node:path";
import * as git from "./git.js";
import { contract, stateDir } from "./paths.js";
import type { Machine } from "./machine.js";
import { checkoutRoot, selectedProjects, workspace, type Manifest, type Project } from "./manifest.js";
import { dumps } from "./jsonmerge.js";
import * as ui from "./ui.js";

const ROOT_FILES = ["CLAUDE.md", "CLAUDE.local.md", ".mcp.json"];
const SKIP_UNDER_CLAUDE = new Set(["worktrees", "plans", "settings.json"]);
const EXCLUDE_LINES = [".claude/", ".mcp.json", "CLAUDE.md", "CLAUDE.local.md"];
const SETTINGS_LOCAL = ".claude/settings.local.json";
const NOT_SYNCED = new Set(["memory", "secrets"]);

export const projectState = (repo: string, p: Project) => join(repo, "projects", p.name);
export const memoryDir = (repo: string, p: Project) => join(projectState(repo, p), "memory");
export function checkouts(p: Project, ws: string): string[] { return checkoutsAt(checkoutRoot(p, ws)); }
/** The checkout at `root` and its worktrees; nothing when the directory is absent. */
function checkoutsAt(root: string): string[] {
  if (!existsSync(root)) return [];
  if (!git.isRepo(root)) return [root];
  const w = git.worktrees(root); return w.length ? w : [root];
}
function walk(dir: string, fn: (f: string) => void, skipDir?: (rel: string) => boolean, base = dir) {
  if (!existsSync(dir)) return;
  for (const e of readdirSync(dir, { withFileTypes: true })) {
    const f = join(dir, e.name); const rel = relative(base, f);
    if (e.isDirectory()) { if (!skipDir?.(rel)) walk(f, fn, skipDir, base); } else if (e.isFile()) fn(rel);
  }
}
function managedRels(base: string): Set<string> {
  const rels = new Set<string>();
  for (const f of ROOT_FILES) if (existsSync(join(base, f)) && statSync(join(base, f)).isFile()) rels.add(f);
  walk(join(base, ".claude"), (rel) => { if (rel !== "settings.json") rels.add(".claude/" + rel); }, (rel) => SKIP_UNDER_CLAUDE.has(rel.split("/")[0]));
  return rels;
}
function sideRels(side: string): Set<string> { const rels = new Set<string>(); walk(side, (rel) => rels.add(rel), (rel) => NOT_SYNCED.has(rel.split("/")[0])); return rels; }
function normalize(rel: string, data: Buffer): Buffer {
  if (rel !== SETTINGS_LOCAL) return data;
  try { const d = JSON.parse(data.toString("utf8") || "{}"); delete d.autoMemoryDirectory; return Buffer.from(Object.keys(d).length ? dumps(d) : ""); } catch { return data; }
}
function localize(rel: string, data: Buffer, mem: string): Buffer {
  if (rel !== SETTINGS_LOCAL) return data;
  let d: any = {}; try { d = data.toString("utf8").trim() ? JSON.parse(data.toString("utf8")) : {}; } catch {}
  d.autoMemoryDirectory = contract(mem); return Buffer.from(dumps(d));
}
/** Which files the project state placed last time, so a file deleted from the state is removed from the checkouts. */
const placedDir = () => join(stateDir(), "project-state");
const stateFile = (p: Project) => join(placedDir(), `${p.name}.json`);
const loadState = (p: Project): Set<string> => { try { return new Set(JSON.parse(readFileSync(stateFile(p), "utf8")).files); } catch { return new Set(); } };
const saveState = (p: Project, files: Set<string>) => { mkdirSync(dirname(stateFile(p)), { recursive: true }); writeFileSync(stateFile(p), JSON.stringify({ files: [...files].sort() }, null, 2)); };
export const dropPlacedRecord = (name: string) => rmSync(join(placedDir(), `${name}.json`), { force: true });
/** The one thing cs wrote into a checkout that points at the share: the auto-memory location in .claude/settings.local.json.
 *  Removes the key; the file goes when nothing else is left in it. Nothing else in the checkout is touched. True when it changed something. */
export function stripPointer(checkout: string): boolean {
  const f = join(checkout, SETTINGS_LOCAL); if (!existsSync(f)) return false;
  let d: any; try { d = JSON.parse(readFileSync(f, "utf8") || "{}"); } catch { return false; }
  if (!("autoMemoryDirectory" in d)) return false;
  delete d.autoMemoryDirectory;
  Object.keys(d).length ? writeFileSync(f, dumps(d)) : unlinkSync(f);
  return true;
}
/** A placed-files record whose name is no longer in the manifest means the project was removed from the share (cs remove
 *  on another machine): strip the pointer from the checkout at the project's default location — plain or `<name>/repo`
 *  with its worktrees — and drop the record. Returns one line per checkout changed. */
export function sweepRemoved(man: Manifest, ws: string): string[] {
  const changes: string[] = []; if (!existsSync(placedDir())) return changes;
  for (const f of readdirSync(placedDir())) {
    if (!f.endsWith(".json")) continue; const name = f.slice(0, -5); if (man.projects[name]) continue;
    const root = existsSync(join(ws, name, "repo", ".git")) ? join(ws, name, "repo") : join(ws, name);
    for (const c of checkoutsAt(root)) if (stripPointer(c)) changes.push(`${name}: auto-memory pointer removed from ${contract(c)} (project removed from the share)`);
    dropPlacedRecord(name);
  }
  return changes;
}
function write(path: string, data: Buffer, mtime?: number) {
  mkdirSync(dirname(path), { recursive: true }); const tmp = path + ".cs-tmp"; writeFileSync(tmp, data);
  if (mtime) utimesSync(tmp, mtime, mtime); renameSync(tmp, path);
}
export function ensureExclude(checkout: string, check: boolean, changes: string[]) {
  if (!existsSync(join(checkout, ".git"))) return;
  const ex = git.infoExclude(checkout); const text = existsSync(ex) ? readFileSync(ex, "utf8") : "";
  const missing = EXCLUDE_LINES.filter((l) => !text.split("\n").includes(l));
  if (!missing.length) return;
  changes.push(`exclude ${missing.join(", ")} in ${contract(checkout)}`);
  if (!check) { mkdirSync(dirname(ex), { recursive: true }); writeFileSync(ex, text + (!text || text.endsWith("\n") ? "" : "\n") + "# claude-share managed files\n" + missing.join("\n") + "\n"); }
}
export function syncProject(repo: string, p: Project, ws: string, check = false): string[] {
  const changes: string[] = []; const side = projectState(repo, p); const targets = checkouts(p, ws);
  if (!targets.length) return changes;
  const mem = memoryDir(repo, p); if (!existsSync(mem) && !check) mkdirSync(mem, { recursive: true });
  const previously = loadState(p);
  const all = new Set<string>(sideRels(side)); for (const t of targets) for (const r of managedRels(t)) all.add(r); all.add(SETTINGS_LOCAL);
  const final = new Set<string>();
  for (const rel of [...all].sort()) {
    const sp = join(side, rel); const sideExists = existsSync(sp) && statSync(sp).isFile();
    const sideData = sideExists ? normalize(rel, readFileSync(sp)) : undefined; const sideMtime = sideExists ? statSync(sp).mtimeMs / 1000 : -1;
    let best = sideData, bestM = sideMtime, from = "project state";
    for (const t of targets) { const tp = join(t, rel); if (existsSync(tp) && statSync(tp).isFile()) { const d = normalize(rel, readFileSync(tp)); const mt = statSync(tp).mtimeMs / 1000;
      if ((!best || !d.equals(best)) && mt > bestM + 1e-6) { best = d; bestM = mt; from = contract(t); } } }
    if (!sideExists && previously.has(rel) && rel !== SETTINGS_LOCAL) {
      for (const t of targets) { const tp = join(t, rel); if (existsSync(tp)) { changes.push(`remove ${rel} from ${contract(t)} (deleted in project state)`); if (!check) unlinkSync(tp); } }
      continue;
    }
    if (best === undefined) { if (rel === SETTINGS_LOCAL) { best = Buffer.alloc(0); bestM = sideMtime; } else continue; }
    if ((!sideData || !best.equals(sideData)) && (best.length || rel !== SETTINGS_LOCAL)) { changes.push(`project state ← ${rel} (from ${from})`); if (!check) write(sp, best, bestM > 0 ? bestM : undefined); }
    if (best.length || rel !== SETTINGS_LOCAL) final.add(rel);
    for (const t of targets) { const tp = join(t, rel); const want = localize(rel, best, mem); const have = existsSync(tp) && statSync(tp).isFile() ? readFileSync(tp) : undefined;
      if (!have || !have.equals(want)) { changes.push(`${contract(t)}/${rel} ← project state`); if (!check) write(tp, want, bestM > 0 ? bestM : undefined); } }
  }
  for (const t of targets) ensureExclude(t, check, changes);
  if (!check) saveState(p, final);
  return changes;
}
export function runLink(repo: string, m: Machine, man: Manifest, names: string[] = [], check = false): number {
  const ws = workspace(man, m);
  const unknown = names.filter((n) => !man.projects[n]); if (unknown.length) throw new Error(`cs: unknown project(s): ${unknown.join(", ")}`);
  let total = 0;
  if (!names.length && !check) for (const c of sweepRemoved(man, ws)) { ui.step(c); total++; }
  for (const p of selectedProjects(man, m)) { if (names.length && !names.includes(p.name)) continue; if (!checkouts(p, ws).length) continue;
    const ch = syncProject(repo, p, ws, check); for (const c of ch) check ? ui.info(`${p.name}: ${c}`) : ui.step(`${p.name}: ${c}`); total += ch.length; }
  if (!total) ui.ok("project files in sync");
  return total;
}
